File: singleProbability.py
# singleProbability.py

import numpy as np
import matplotlib.pyplot as plt
from model import sample_parameters, infection_probability, occupancy_params
from randomManager import RandomNumberManager


def parse_args():
    import argparse

    parser = argparse.ArgumentParser(description="Run single occupancy category probability simulation.")
    parser.add_argument('--N', type=int, default=10000, help="Number of simulations (default 10000)")
    parser.add_argument('--category', type=str, default='Classroom', choices=occupancy_params.keys(), help="Occupancy category")
    parser.add_argument('--community_rate', type=float, default=None, help="Community infection rate (0–1), optional")
    parser.add_argument('--no_zero_infectors', action='store_false', dest='allow_zero_infectors', help="Disallow zero infected simulations (default allows zero infectors)")
    parser.add_argument('--ecai', type=float, default=None, help="ECAi in L/s/person (default from occupancy params)")
    parser.add_argument('--show_plots', action='store_true', help="Show plots of probability distribution")

    return parser.parse_args()


def print_progress(iteration, total):
    bar_length = 100
    progress = int((iteration / total) * bar_length)
    bar = '*' * progress + ' ' * (bar_length - progress)
    print(f"\rProgress: |{bar}| {iteration}/{total}", end='', flush=True)


def main():
    args = parse_args()

    N = args.N
    category = args.category
    community_rate = args.community_rate
    allow_zero_infectors = args.allow_zero_infectors
    ECAi = args.ecai if args.ecai is not None else occupancy_params[category]["ECAi"]

    rng = RandomNumberManager()
    probabilities = []
    zero_infectors_count = 0

    for i in range(1, N + 1):
        par = sample_parameters(rng, category=category)
        comm_rate = 0 if community_rate is None else community_rate
        prob, infected_flag = infection_probability(
            ECAi,
            par,
            rng,
            category=category,
            require_infectors=not allow_zero_infectors,
            override_community_rate=comm_rate
        )
        probabilities.append(prob)
        if infected_flag == 0:
            zero_infectors_count += 1

        # Update progress
        if N >= 100 and i % (N // 100) == 0:
            print_progress(i, N)

    print_progress(N, N)
    print("\n")  # Newline after progress bar

    probabilities = np.array(probabilities) * 100  # Convert to %
    p96 = np.percentile(probabilities, 96)
    zero_percent = (zero_infectors_count / N) * 100

    print(f"\nSimulation results for category '{category}':")
    print(f"  Simulations: {N}")
    print(f"  ECAi: {ECAi} L/s/person")
    if community_rate is None:
        default_rate = occupancy_params[category]["community_rate"]
        print(f"  Community infection rate: Default ({default_rate})")
    else:
        print(f"  Community infection rate: {community_rate}")
    print(f"  Allow zero infectors: {allow_zero_infectors}")
    print(f"  96th percentile of infection probability: {p96:.3f}%")
    print(f"  Percentage of simulations with zero infectors: {zero_percent:.1f}%\n")

    if args.show_plots:
        fig, axs = plt.subplots(2, 1, figsize=(10, 10))

        # Plot: Linear scale
        axs[0].hist(probabilities, bins=1000, color='steelblue', edgecolor='black')
        axs[0].set_xlabel('Infection Probability (%)')
        axs[0].set_ylabel('Count')
        axs[0].set_title(f'Probability Distribution (Linear Scale) - {category}')
        axs[0].grid(True, axis='y', linestyle='--', linewidth=0.5)

        # Plot: Log10-transformed data on linear scale
        probabilities_nozero = probabilities[probabilities > 0.001]
        log_probs = np.log10(probabilities_nozero)

        axs[1].hist(log_probs, bins=100, color='steelblue', edgecolor='black')
        axs[1].set_xlabel('log₁₀(Infection Probability [%])')
        axs[1].set_ylabel('Count')
        axs[1].set_title(f'Probability Distribution (Log10-Transformed) - {category}')
        axs[1].grid(True, axis='y', linestyle='--', linewidth=0.5)

        plt.tight_layout()
        plt.show()


if __name__ == "__main__":
    main()

File: C++/src/single_probability.cpp
// single_probability.cpp — port of singleProbability.py:
// Run infection probability simulations for a single occupancy category

#include "model.h"
#include "random_manager.h"

#include <vector>
#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <map>
#include <cmath>

int main(int argc, char* argv[]) {
    int N = 10000;
    std::string category = "Classroom";
    double community_rate = -1;  // -1 means use default
    bool allow_zero_infectors = true;
    QERDistribution qer_distribution{};
    double ecai_override = -1;   // -1 means use default from params

    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--N" && i + 1 < argc) {
            N = std::atoi(argv[++i]);
        } else if (arg == "--category" && i + 1 < argc) {
            category = argv[++i];
        } else if (arg == "--community_rate" && i + 1 < argc) {
            community_rate = std::atof(argv[++i]);
        } else if (arg == "--ecai" && i + 1 < argc) {
            ecai_override = std::atof(argv[++i]);
        } else if (arg == "--no_zero_infectors") {
            allow_zero_infectors = false;
        } else if (arg == "--use-fitted-qer") {
            qer_distribution = {QERDistributionKind::JonesFitted, {}};
        } else if (arg == "--qer-distribution" && i + 1 < argc) {
            if (!parse_qer_distribution(argv[++i], qer_distribution)) {
                fprintf(stderr, "Unknown QER distribution. See README for valid profile names.\n");
                return 2;
            }
        } else if (arg == "--show_plots") {
            // Ignored — no plotting in C++
        } else if (arg == "--help" || arg == "-h") {
            printf("Usage: single_probability [options]\n");
            printf("  --N <int>            Number of simulations (default 10000)\n");
            printf("  --category <name>    Occupancy category (default Classroom)\n");
            printf("  --community_rate <f> Community infection rate 0-1 (default: from params)\n");
            printf("  --ecai <float>       ECAi in L/s/person (default: from params)\n");
            printf("  --no_zero_infectors  Disallow zero infector simulations\n");
            printf("  --qer-distribution   jones, jones-fitted, or one of the 30\n");
            printf("                        mikszewski-<organism>-<activity> profiles\n");
            printf("                        (default jones)\n");
            printf("  --use-fitted-qer     Alias for --qer-distribution fitted\n");
            return 0;
        }
    }

    if (N < 1) { printf("Invalid N, using default 10000.\n"); N = 10000; }

    auto it = occupancy_params.find(category);
    if (it == occupancy_params.end()) {
        printf("Error: '%s' is not a valid category.\n", category.c_str());
        printf("Available categories:");
        for (const auto& k : category_order)
            printf(" %s", k.c_str());
        printf("\n");
        return 1;
    }

    double ECAi = (ecai_override >= 0) ? ecai_override : it->second.ECAi;
    double comm_rate = (community_rate >= 0) ? community_rate : 0;  // 0 means use default
    bool use_default_comm = (community_rate < 0);

    RandomNumberManager rng;
    std::vector<double> probabilities;
    probabilities.reserve(N);
    int zero_infectors_count = 0;

    for (int i = 0; i < N; i++) {
        auto par = sample_parameters(rng, category);
        double cr = use_default_comm ? 0 : comm_rate;
        auto [prob, infected_flag] = infection_probability(
            ECAi, par, rng, category, !allow_zero_infectors, cr,
 qer_distribution);
        probabilities.push_back(prob);
        if (infected_flag == 0) zero_infectors_count++;
    }

    std::sort(probabilities.begin(), probabilities.end());
    double rank = 0.96 * (N - 1);
    int lo = static_cast<int>(std::floor(rank));
    int hi = static_cast<int>(std::ceil(rank));
    double frac = rank - lo;
    double p96 = (probabilities[lo] * (1 - frac) + probabilities[hi] * frac) * 100;
    double zero_percent = 100.0 * zero_infectors_count / N;

    printf("\nSimulation results for category '%s':\n", category.c_str());
    printf("  Simulations: %d\n", N);
    printf("  ECAi: %g L/s/person\n", ECAi);
    if (use_default_comm) {
        printf("  Community infection rate: Default (%.4f)\n", it->second.community_rate);
    } else {
        printf("  Community infection rate: %.4f\n", comm_rate);
    }
    printf("  Allow zero infectors: %s\n", allow_zero_infectors ? "true" : "false");
    printf("  96th percentile of infection probability: %.3f%%\n", p96);
    printf("  Percentage of simulations with zero infectors: %.1f%%\n\n", zero_percent);

    return 0;
}